File: DFS/022.Generate-Parentheses/022.Generate-Parentheses.java
class Solution {
    List<String> ret;
    int n;
    public List<String> generateParenthesis(int n) {
        ret = new ArrayList<>();
        this.n = n;
        dfs("", 0, 0);
        return ret;
    }
    private void dfs(String current, int left, int right) {
        if (current.length() == n * 2) {
            if (left == right) {
                ret.add(current);
            }
            return;
        }

        if (left < n) {
            dfs(current + "(", left + 1, right);
        }
        if (right < left) {
            dfs(current + ")", left, right + 1);
        }
    }
}

File: DFS/022.Generate-Parentheses/022.Generate-Parentheses.ts
function generateParenthesis(n: number): string[] {
  const ret: string[] = [];
  backTrace('', 0, 0);
  return ret;
  function backTrace(current: string, left: number, right: number) {
    if (current.length === n * 2) {
      if (left === right) {
        ret.push(current);
      }
      return;
    }
    if (left < n) {
      backTrace(`${current}(`, left + 1, right);
    }
    if (left > right) {
      backTrace(`${current})`, left, right + 1);
    }
  }
}
